### PR TITLE
Fix being unable to pick up latest non-empty cart started in shop and continued in API

### LINKED ISSUE #14539: API tokenValue may be missing on new cart
**Sylius version affected**: 1.11 / 1.12 / master

**Description**  

Creating a cart on the website does not create a tokenValue besides it's used as ID with the API. When I try to get an already created cart using the API it does not return the tokenValue.

**Steps to reproduce**  

1. Use the sylius website, login, and add 1 item to cart
2. Use the sylius API, login and use the token for further requests
3. Use the endpoint POST `/v2/shop/orders` to create a cart
4. The cart has no tokenValue and it's not possible to use it with the API

**Possible Solution**  

Generate the tokenValue before inserting the order in the database. (I assume this has consequences that I'm not aware of)

### PR DESCRIPTION
**Pull request by jakubtobiasz**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.12                  |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| Related tickets | fixes #14539, successor of #14597                      |
| License         | MIT                                                          |

PR created by @Nek- was great, but was involving some sort of changes we've called feature and bc policy breaking. We (believe that) we've found a way to fix the bug part without changing the previous behavior on the web shop.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/config/services/factories.xml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <services>
+        <defaults public="true" />
+
+        <service id="sylius_core.factory.order" class="Sylius\Bundle\CoreBundle\Factory\OrderFactory" decorates="sylius.factory.order">
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services/factories.xml]:
```suggestion
        <service id="sylius.custom_factory.order" class="Sylius\Bundle\CoreBundle\Factory\OrderFactory" decorates="sylius.factory.order">
```
To be in line with other factory decorations. [ref](https://github.com/Sylius/Sylius/blob/fb70c5e34a5e246c6a5426ecd182d17f73efb808/src/Sylius/Bundle/CoreBundle/Resources/config/services.xml#L149)

**Review by NoResponseMate — APPROVED** [review]

File: src/Sylius/Behat/Resources/config/suites/composite/cart/shopping_cart.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# This file is part of the Sylius package.
+# (c) Paweł Jędrzejewski
+
+default:
+    suites:
+        api_shopping_cart:
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Resources/config/suites/composite/cart/shopping_cart.yml]:
```suggestion
        composite_shopping_cart:
```

File: features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    In order to order an item
+    As a Customer
+    I want to be able to make an order on UI and resume on API
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "Apple's Polishing Cloth"
+        And I am a logged in customer on the web store and in the API
+
+    @api @composite
```
Comment (NoResponseMate) [inline features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature]:
```suggestion
    @composite
```

**Review by NoResponseMate — APPROVED** [review]

File: features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature [hunk 4]
```diff
@@ -0,0 +1,16 @@
+@shopping_cart
+Feature: Continuing processing with using API of the  cart started on the web store
```
Comment (GSadee) [inline features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature]:
```suggestion
Feature: Continuing to process the cart picked up in the web store, in the API
```

File: features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    In order to order an item
```
Comment (GSadee) [inline features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature]:
```suggestion
    In order to order an item easily
```
🤔

File: features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    As a Customer
+    I want to be able to make an order on UI and resume on API
```
Comment (GSadee) [inline features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature]:
```suggestion
    I want to be able to make an order in the UI and resume in the API
```

File: features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature (hunk 3 above)
Comment (GSadee) [reply features/cart/shopping_cart/continuing_cart_processing_from_ui_in_api.feature]:
```suggestion
    @api-and-ui
```
Perhaps something more descriptive? 🤔

File: src/Sylius/Behat/Context/Composite/Shop/CompositeCartContext.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+use Sylius\Behat\Context\Api\Shop\CartContext;
+use Sylius\Behat\Context\Ui\Shop\CartContext as UiCartContext;
+use Sylius\Component\Core\Model\ProductInterface;
+
+class CompositeCartContext implements Context
+{
+    public function __construct(
+        private CartContext $apiCartContext,
+        private UiCartContext $uiCartContext,
+    ) {}
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Composite/Shop/CompositeCartContext.php]:
```suggestion
    ) {
    }
```

File: src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+use Sylius\Behat\Context\Setup\ShopSecurityContext;
+use Sylius\Behat\Service\SharedStorageInterface;
+
+class CompositeCustomerContext implements Context
+{
+    public function __construct(
+        private ShopSecurityContext $uiSecurityContext,
+        private ShopSecurityContext $apiSecurityContext,
+        private SharedStorageInterface $sharedStorage,
+    ) {}
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php]:
```suggestion
    ) {
    }
```

File: src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * @Then /^I am a logged in customer on the web store and in the API$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php]:
```suggestion
     * @Given I am a logged in customer on the web store and in the API
```

File: src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+     */
+    public function IAmALoggedInCustomerOnTheApiAndTheUi()
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php]: Missing return type

File: src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Behat\Context\Composite\Shop;
+
+use Behat\Behat\Context\Context;
+use Sylius\Behat\Context\Setup\ShopSecurityContext;
+use Sylius\Behat\Service\SharedStorageInterface;
+
+class CompositeCustomerContext implements Context
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php]:
```suggestion
class CompositeShopSecurityContext implements Context
```

File: src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Behat\Context\Composite\Shop;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Composite/Shop/CompositeCustomerContext.php]:
```suggestion
namespace Sylius\Behat\Context\Composite\Setup;
```

File: src/Sylius/Bundle/CoreBundle/spec/Factory/OrderFactorySpec.php [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+    public function it_creates_an_order(FactoryInterface $baseFactory): void
+    {
+        $someOrder = new Order();
+        $baseFactory->createNew()->willReturn($someOrder);
+
+        $this->createNew()->shouldReturn($someOrder);
+    }
+
+    public function it_creates_a_cart(FactoryInterface $baseFactory): void {
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/Factory/OrderFactorySpec.php]:
```suggestion
    public function it_creates_a_cart(FactoryInterface $baseFactory): void
    {
```

File: src/Sylius/Bundle/CoreBundle/spec/Factory/OrderFactorySpec.php [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+        $this->beConstructedWith($baseFactory);
+    }
+
+    public function it_creates_an_order(FactoryInterface $baseFactory): void
+    {
+        $someOrder = new Order();
+        $baseFactory->createNew()->willReturn($someOrder);
+
+        $this->createNew()->shouldReturn($someOrder);
+    }
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/Factory/OrderFactorySpec.php]: I would be for mocking an order by its interface in the arguments of the spec function, the same applies to the below scenario

File: src/Sylius/Bundle/CoreBundle/spec/Factory/OrderFactorySpec.php [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        $order->setChannel($channel);
+        $order->setLocaleCode('en_US')->shouldBeCalled();
+        $order->setCurrencyCode('USD')->shouldBeCalled();
+
+        $currency->getCode()->willReturn('USD');
+        $channel->getBaseCurrency()->willReturn($currency);
+
+        $cart = $this->createNewCart(
+            $channel->getWrappedObject(),
+            null,
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/Factory/OrderFactorySpec.php]: As the spec is for creating the cart with the customer, you need to pass here a customer and check setting them

**Comment by Nek-** [comment]
The most important change regarding my PR is that you decided that an order must not always have a tokenValue.

Why?

**Comment by jakubtobiasz** [comment]
@Nek- our decision was to keep the old behavior and fix the bug you reported. The main reason is safety – we fixed the root of the problem and didn't change the behavior of the shop cart.

File: src/Sylius/Behat/Resources/config/services/contexts/hybrid/setup.xml [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd"
+>
+    <services>
+        <defaults public="true" />
+
+        <service id="sylius.behat.context.hybrid.shop.composite_cart" class="Sylius\Behat\Context\Hybrid\Setup\CompositeCartContext">
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Resources/config/services/contexts/hybrid/setup.xml]:
```suggestion
        <service id="sylius.behat.context.hybrid.shop.cart" class="Sylius\Behat\Context\Hybrid\Setup\CartContext">
```
No point in keeping the `composite` part, the same below

File: src/Sylius/Behat/Resources/config/services/contexts.xml [hunk 17]
```diff
@@ -15,6 +15,7 @@
     <imports>
         <import resource="contexts/api.xml" />
         <import resource="contexts/cli.xml" />
+        <import resource="contexts/hybrid.xml" />
         <import resource="contexts/domain.xml" />
         <import resource="contexts/hook.xml" />
         <import resource="contexts/setup.xml" />
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Resources/config/services/contexts.xml]:
```suggestion
        <import resource="contexts/api.xml" />
        <import resource="contexts/cli.xml" />
        <import resource="contexts/domain.xml" />
        <import resource="contexts/hook.xml" />
        <import resource="contexts/hybrid.xml" />
        <import resource="contexts/setup.xml" />
```

File: src/Sylius/Behat/Context/Hybrid/Setup/CompositeCartContext.php [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+use Sylius\Behat\Context\Api\Shop\CartContext;
+use Sylius\Behat\Context\Ui\Shop\CartContext as UiCartContext;
+use Sylius\Component\Core\Model\ProductInterface;
+
+class CompositeCartContext implements Context
+{
+    public function __construct(
+        private CartContext $apiCartContext,
+        private UiCartContext $uiCartContext,
+    ) {}
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Hybrid/Setup/CompositeCartContext.php]:
```suggestion
    ) {
    }
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Jakub! :1st_place_medal:

**Pull request merged**